Carbon.Windows.Installer/Functions/Get-Msi.ps1

```diff
@@ -0,0 +1,310 @@
+                    continue
+                }
+            }
+            catch
+            {
+                $msg = "Exception opening MSI database in file ""$($msiPath)"": $($_)"
+                Write-Error -Message $msg -ErrorAction $ErrorActionPreference
+                continue
+            }
+
+            try
+            {
+                Debug '    _Tables'
+                $tables = Read-MsiTable -Database $database -Name '_Tables' -MsiPath $msiPath
+                $info.TableNames = $tables | Select-Object -ExpandProperty 'Name'
+
+                foreach( $tableName in $info.TableNames )
+                {
+                    $info.Tables | Add-Member -Name $tableName -MemberType NoteProperty -Value @()
+                    if( $IncludeTable -and -not ($IncludeTable | Where-Object { $tableName -like $_ }) )
+                    {
+                        Debug "  ! $($tableName)"
+                        continue
+                    }
+
+                    Debug "    $($tableName)"
+                    $info.Tables.$tableName = Read-MsiTable -Database $database -Name $tableName -MsiPath $msiPath
+                }
+
+                [Guid] $productCode = [Guid]::Empty
+                [String] $rawProductCode = $info.GetPropertyValue('ProductCode')
+                if( [Guid]::TryParse($rawProductCode, [ref]$productCode) )
+                {
+                    $info.ProductCode = $productCode
+                }
+
+                [int] $langID = 0
+                [String] $rawLangID = $info.GetPropertyValue('ProductLanguage')
+                if( [int]::TryParse($rawLangID, [ref]$langID) )
+                {
+                    $info.ProductLanguage = $langID
+                }
+
+                $info.Path = $msiPath;
+                $info.Manufacturer = $info.GetPropertyValue('Manufacturer')
+                $info.ProductName = $info.GetPropertyValue('ProductName')
+                $info.ProductVersion = $info.GetPropertyValue('ProductVersion')
+
+                [void]$info.pstypenames.Insert(0, 'Carbon.Windows.Installer.MsiInfo')
+            }
+            finally
+            {
+                $collect = $false
+                if( $database )
+                {
+                    [void][Runtime.InteropServices.Marshal]::ReleaseCOMObject($database)
+                    $collect = $true
+                }
+
+                if( $installer )
+                {
+                    [void][Runtime.InteropServices.Marshal]::ReleaseCOMObject($installer)
+                    $collect = $true
+                }
+
+                if( $collect )
+                {
+                    # ReleaseCOMObject still leaves the MSI file open. The only way to close the file handle is to run
+                    # garbage collection, and even then it takes a few seconds. :(
+                    Debug "[GC]::Collect()  START"
+                    [GC]::Collect()

```

Comment:
Hopefully this doesn't cause any unintended side effects in the rest of the code running in the current PowerShell session.

Reply:
It shouldn't. At worse, the function will take a little longer while the garbage collector runs.